### PR TITLE
Update developer documentation to reflect upcoming 2.8 changes

### LINKED ISSUE #861: Update developer documentation ahead of 2.8 release
We've changed enough that it's fallen out of date.

### PR DESCRIPTION
**Pull request by eslavich**
Here are some developer overview updates to reflect changes that we've merged for 2.8.  Let me know if I'm missing any topics you'd like to see or if I should expand more on any of what I have.

Resolves #861

### CODE REVIEW

File: docs/asdf/developer_overview.rst [hunk 1]
```diff
@@ -34,11 +34,50 @@ Some terminology and definitions
...
+on the top-level ``asdf`` module.
+
+**Entry point:** A Python packaging feature that allows a given package
+to use plugins registered by other packages.  See [https://packaging.python.org/specifications/entry-points/]
+for more information.
+
+**Resource mapping:** An asdf plugin that provides access to "resources" which
+are binary blobs associated with a URI.  These resources are mostly schemas,
+but any resource may provided by a mapping.  Resource mappings are provided
+via entry points or added at runtime using a method on the global config object.
```
Comment (perrygreenfield) [inline docs/asdf/developer_overview.rst:59]: And as a replacement for the previous resolver mechanism.

File: docs/asdf/developer_versioning.rst [hunk 2]
```diff
@@ -255,8 +257,8 @@ The new AsdfType might look something like this:
         version = "2.0.0"
 
 **CAUTION:** We might be tempted here to simply update the original
-BreakingObjectType, but failing to handle an older version of the schema
+BreakingObjectType, but failing to handle an older version of the tag
 constitutes dropping support for any ASDF Standard version that relies
-on that schema.  This should only be done after a deprecation period and
+on that tag.  This should only be done after a deprecation period and
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_versioning.rst:262]: And a compendium of keyboard jokes? Forget I wrote that.

File: docs/asdf/developer_overview.rst [hunk 3]
```diff
@@ -34,11 +34,50 @@ Some terminology and definitions
...
 protocol, though for asdf, this isn't necessarily the case as mentioned next.
+Recent changes to the library permit use of URIs with the asdf:// scheme, which
+is intended to reduce confusion over the distinction between identifiers
+and locations.
 
 **Resolver:** Tools to map URIs and tags into actual locations of schema files,
 which may be local directories (the usual approach) or an actual URL for
 retrieval over the network. This is more complicated that it may seem for
-reasons explained later.
+reasons explained later.  The resolver system has been deprecated in favor
```
[the remark this answers is not in the document]
Comment (eslavich) [reply docs/asdf/developer_overview.rst:90]: Added an early section on the redesign.

File: docs/asdf/developer_overview.rst [hunk 4]
```diff
@@ -565,4 +634,27 @@ with these transformed children, and that is the node that gets handed to
...
+
+The new ``Converter`` interface also accepts a ``ctx`` variable, but
+instead of an ``AsdfFile`` it's an instance of ``SerializationContext``.  This
+new object will serve the purpose of configuring serialization parameters
+and keeping necessary state, which means that the ``AsdfFile`` can go
+unmodified.  The ``SerializationContext`` will be relatively lightweight and
+creating it will not incur as much of a performance penalty as creating an
+``AsdfFile``.
+
 Thus reader, your mind shall now be drained.
```
[the remark this answers is not in the document]
Comment (eslavich) [reply docs/asdf/developer_overview.rst]: Removed.

File: docs/asdf/developer_overview.rst [hunk 5]
```diff
@@ -34,11 +34,50 @@ Some terminology and definitions
...
 retrieval over the network. This is more complicated that it may seem for
-reasons explained later.
+reasons explained later.  The resolver system has been deprecated in favor
+of resource mappings; new code should use the latter instead.
+
+**Global config:** A global library configuration feature that was added in
+asdf 2.8.  Allows plugins to be added or removed at runtime and ``AsdfFile``
+defaults to be modified by the user.  Accessed via the ``get_config`` method
+on the top-level ``asdf`` module.
+
```
[the remark this answers is not in the document]
Comment (eslavich) [reply docs/asdf/developer_overview.rst:102]: I added a couple of examples here.

File: docs/asdf/developer_overview.rst [hunk 6]
```diff
@@ -34,11 +34,50 @@ Some terminology and definitions
...
+**Entry point:** A Python packaging feature that allows a given package
+to use plugins registered by other packages.  See [https://packaging.python.org/specifications/entry-points/]
+for more information.
```
[the remark this answers is not in the document]
Comment (eslavich) [reply docs/asdf/developer_overview.rst:109]: Extended this section accordingly.

File: docs/asdf/developer_overview.rst (hunk 1 above)
Comment (eslavich) [reply docs/asdf/developer_overview.rst:114]: Added a sentence to that effect.

File: docs/asdf/developer_overview.rst [hunk 7]
```diff
@@ -67,43 +106,71 @@ aspect to this since schema should be language agnostic and in that view, not
...
+and ``ResourceMappingProxy``, respectively) that carry additional metadata
+like the package name and version of the entry point, and add some convenience
+methods on top of what the extension developer provides.  Additionally, ``ExtensionProxy``
+allows the library to treat both new-style ``Extension`` instances and old-style
+``AsdfExtension`` instances similarly.
+
+To see the list of extensions loaded by the library, call ``asdf.get_config().extensions``.
+To see the list of resource mappings, call ``asdf.get_config().resource_mappings``.
+Both of these properties are lazy-loaded and then cached, so the first call will take
+a moment to complete but subsequent calls will return immediately.
```
[the remark this answers is not in the document]
Comment (eslavich) [reply docs/asdf/developer_overview.rst:187]: I haven't written any user documentation for the new interfaces, and now that asdf-astropy is shaping up it's probably getting to be time to do that.  I'll add an issue to tackle that task separately?

[the remark this answers is not in the document]
Comment (eslavich) [reply docs/asdf/developer_overview.rst:187]: I forgot that we still needed to provide user documentation for the new interfaces.  Added an issue for that: https://github.com/asdf-format/asdf/issues/880

**Comment by eslavich** [comment]
@perrygreenfield I made the necessary updates, except for the bit about troubleshooting resource mappings which I turned into a new issue.  Also I'm fresh out of keyboard jokes.

**Review by perrygreenfield — APPROVED** [review]
Looks great

**Pull request merged**